File: Data Structures & Algorithms/climbing-stairs/submission-1.java
class Solution {
    public int climbStairs(int n) {
        int[] dp=new int[n+1];
        int res=memoization(dp,n);
        return res;
    }
    public int memoization(int[] dp, int n){
        if(dp[n]!=0){
            return dp[n];
        }
        if(n<0){
            return 0;
        }
        if(n<3){
            return n;
        }
        dp[n]=memoization(dp,n-1)+memoization(dp,n-2);
        return dp[n];
    }
}

File: Data Structures & Algorithms/permutations/submission-1.py
class Solution:
    def permute(self, nums: List[int]) -> List[List[int]]:
        self.res = []
        used = [False] * len(nums)
        self.backtrack([], nums, used)
        return self.res
    
    def backtrack(self, perm: List[int], nums: List[int], used: List[bool]):
        if len(perm) == len(nums):
            self.res.append(perm[:])
            return
        
        for i in range(len(nums)):
            if not used[i]:
                perm.append(nums[i])
                used[i] = True
                self.backtrack(perm,nums,used)
                perm.pop()
                used[i] = False
        
